Add support for Jackson serialization using LZ4 compression #27066

diversit opened the pull request:
<!--
# Pull Request Checklist

* [ ] Have you read through the [contributor guidelines](./CONTRIBUTING.md)?
* [ ] Have you signed the [Lightbend CLA](https://www.lightbend.com/contribute/cla)?
* [ ] Have you updated the documentation?
* [ ] Have you added tests for any changed functionality?
-->
## Purpose

Adds support for LZ4 compression for Jackson json serialisation.

## References

References #27066

## Changes

* added 'algorithm' configuration setting supporting both gzip (by default) or lz4.
* Added lz4-java dependency to akka-serialization-jackson module.
* Updated JacksonSerializer compress and decompress methods to support both compression algorithms. Moved check for 'compressLargerThan' to compress method and 'isGZipped' to decompress method.
* Added test cases for gzip and LZ4 for both Java and Scala objects.
* updated serialization-jackson docs


## Background Context

Could not find a way to easily detect whether input is compressed with LZ4 since it does not have a 'magic' header. So both sending and receiving end must have the compression algorithm set to LZ4. Also, the optional compression when message > minumum-size does not work in this case, so it's compressing in LZ4 regardless of the size. This is added to the documentation.
The LZ4 fast deserialiser requires then uncompressed size to be know. Since that info is not available when receiving a compressed message, I had to use the deserialise-with-unknown-size feature.
The same Input-/OutputStream compression/decompression logic is used as for Gzip only with LZ4 implementations.

diversit committed:
Add support for Jackson serialization using LZ4 compression #27066

* added 'algorithm' configuration setting supporting both gzip (by default) or lz4.
* Added lz4-java dependency to akka-serialization-jackson module.
* Updated JacksonSerializer compress and decompress methods to support both compression algorithms. Moved check for 'compressLargerThan' to compress method and 'isGZipped' to decompress method.
* Added test cases for gzip and LZ4 for both Java and Scala objects.
* updated serialization-jackson docs

patriknw commented on akka-serialization-jackson/src/main/resources/reference.conf:
```diff
@@ -24,9 +24,17 @@ akka.serialization.jackson {
+akka.serialization.jackson {
+  # gzip - using common java gzip
+  # lz4 - LZ4 compression.
+  algorithm = gzip
```
more specifc: `compression-algorithm`

but it's better if we add a subsection for `compression`

```
akka.serialization.jackson.compression {
  algorithm = gzip
  compress-larger-than = 32 KiB
}
```

diversit replied:
Updated to use `akka.serialization.jackson.compression`.

patriknw commented on akka-serialization-jackson/src/main/resources/reference.conf:
```diff
@@ -24,9 +24,17 @@ akka.serialization.jackson {
 akka.serialization.jackson {
-  # The serializer will compress the payload when it's larger than this value.
+  # For GZIP Only: The serializer will compress the payload when it's larger than this value.
   # Compression can be disabled with value 'off'.
```
let's support `algorithm = off` to disable compression completely instead of having that on the `compress-larger-than`

diversit replied:
Added algorithm option 'off' and removed 'off' feature for `compress-larger-than`.

patriknw commented on akka-serialization-jackson/src/main/scala/akka/serialization/jackson/JacksonSerializer.scala:
```diff
@@ -372,29 +383,45 @@ import com.fasterxml.jackson.dataformat.smile.SmileFactory
+        case Compression.LZ4 => new LZ4FrameInputStream(bais)
+      }
+
+      //    val in = new GZIPInputStream(new ByteArrayInputStream(bytes))
```
remove

diversit replied:
Removed.

patriknw commented on project/Dependencies.scala:
```diff
@@ -99,6 +99,7 @@ object Dependencies {
     val jacksonAfterburner = "com.fasterxml.jackson.module" % "jackson-module-afterburner" % jacksonVersion // ApacheV2
     val jacksonCbor = "com.fasterxml.jackson.dataformat" % "jackson-dataformat-cbor" % jacksonVersion // ApacheV2
     val jacksonSmile = "com.fasterxml.jackson.dataformat" % "jackson-dataformat-smile" % jacksonVersion // ApacheV2
+    val lz4_java = "org.lz4" % "lz4-java" % "1.6.0"
```
rename to `lz4Java` or just `lz4`

patriknw replied:
add `// ApacheV2` note at the end

patriknw replied:
It has zero transitive dependencies, good.
One consideration is if we should make lz4-java an optional dependency, but on the other hand if compression is to be used there it's better to use lz4 than gzip. Maybe we should make lz4 the default compression algorithm.
I'm thinking that perhaps we should make the compressed versions separate serializers so we have bindings:
* `jackson-json` - no compression
* `jackson-cbor` - no compression
* `jackson-json-lz4` - lz4 compression
* `jackson-cbor-lz4` - lz4 compression

For gzip it can optionally be enabled for the plain `jackson-json` since it has the magic (we mostly need to support gzip for backwards compatibility with Lagom).

I could do this part as a separate PR afterwards if you like.

patriknw replied:
Or we could add our own magic prefix for lz4. Perhaps that is better?

diversit replied:
Renamed to `lz4Java` and added Apache license note.
I cannot oversee the impact of changing the default compression to lz4. Would that not be a breaking change? Could it be something would break when doing a rolling update or when communicating with external services or db?
Separate serialisers seems a good idea. Would you do that? I think you'd be way faster than me.

Adding own magic prefix would not be compatible with non-akka services sending or expecting a normal LZ4 byte stream, so I would not recommend that.

patriknw replied:
Looks like there is a magic in the specification, and lz4-java has support for it: https://github.com/lz4/lz4-java/blob/master/src/java/net/jpountz/lz4/LZ4FrameOutputStream.java

diversit replied:
Mmm ... indeed. It was not mentioned in the LZ4 spec, but I'll have a look.

patriknw reviewed (commented):
Excellent, thanks for picking up this feature.
A few suggestions...

patriknw commented:
I'll run some jmh benchmarks next week to see how it compares.

patriknw commented:
OK TO TEST

diversit committed:
Moved settings to 'compression' subsection.
Added algorithm 'off' option
Removed 'off' option for 'compress-larger-than'
Updated documentation.
Added test case for compression off.
Renamed lz4-java dependency and added license note.

diversit committed:
Merge pull request #1 from akka/master

update base

diversit commented:
Resolved conflict by rebasing on updated master.

diversit committed:
Revert arrows to normal arrow instead of ascii arrow. Not changed intentionally. Intellij copy magic I presume.

patriknw commented on akka-serialization-jackson/src/test/scala/akka/serialization/jackson/JacksonSerializerSpec.scala:
```diff
@@ -384,6 +384,40 @@ abstract class JacksonSerializerSpec(serializerName: String)
     import JavaTestMessages._
 
+    "serialize with gzip" in withSystem("""
+        akka.serialization.jackson {
```
those should now be `akka.serialization.jackson.compression`

patriknw commented on akka-serialization-jackson/src/main/scala/akka/serialization/jackson/JacksonSerializer.scala:
```diff
@@ -357,29 +360,48 @@ import com.fasterxml.jackson.dataformat.cbor.CBORFactory
+      val compressedOut = compress(new ByteArrayOutputStream(BufferSize))
+      try compressedOut.write(bytes)
+      finally compressedOut.close()
+      new ByteArrayOutputStream(BufferSize).toByteArray
```
This is wrong, returning empty byte array.
```
      val bos = new ByteArrayOutputStream(BufferSize)
      val compressedOut = compress(bos)
      try compressedOut.write(bytes)
      finally compressedOut.close()
      bos.toByteArray
```

patriknw commented on akka-serialization-jackson/src/main/scala/akka/serialization/jackson/JacksonSerializer.scala:
```diff
@@ -357,29 +360,48 @@ import com.fasterxml.jackson.dataformat.cbor.CBORFactory
+      case Compression.OFF => bytes
+      case Compression.GZip if bytes.length <= compressLargerThan => bytes
+      case Compression.GZip => compressWith(new GZIPOutputStream(_))
+      case Compression.LZ4 => compressWith(new LZ4FrameOutputStream(_))
```
Use `LZ4FrameOutputStream.BLOCKSIZE.SIZE_64KB` parameter.

JMH results are horrible with the default 4MB. I tried with same test data as in https://github.com/akka/akka/pull/27067

Size in bytes:
```
off:  5360
gzip: 3469
lz4:  4215
```

JMH results for `JacksonSerializationBench` with that data
```
[info] Benchmark                                                         (algo)  (serializerName)   Mode  Cnt         Score        Error   Units
[info] JacksonSerializationBench.large                                      off      jackson-json  thrpt    5     13480.766 ±   1889.994   ops/s
[info] JacksonSerializationBench.large:·gc.alloc.rate.norm                  off      jackson-json  thrpt    5     77881.103 ±     13.115    B/op

[info] JacksonSerializationBench.large                                     gzip      jackson-json  thrpt    5      4552.216 ±    310.308   ops/s
[info] JacksonSerializationBench.large:·gc.alloc.rate.norm                 gzip      jackson-json  thrpt    5    120508.622 ±     37.079    B/op

[info] JacksonSerializationBench.large                                      lz4      jackson-json  thrpt    5       454.367 ±     33.257   ops/s
[info] JacksonSerializationBench.large:·gc.alloc.rate.norm                  lz4      jackson-json  thrpt    5  16911932.851 ±    364.928    B/op
```

Switching to 64KB it's looking better

```
[info] Benchmark                                                     (algo)  (serializerName)   Mode  Cnt       Score       Error   Units
[info] JacksonSerializationBench.large                                  lz4      jackson-json  thrpt    5    9322.556 ±   992.144   ops/s
[info] JacksonSerializationBench.large:·gc.alloc.rate.norm              lz4      jackson-json  thrpt    5  380593.746 ±    17.155    B/op
```

Still generating pretty much garbage. Might be something more that can be tuned?

patriknw reviewed (commented):
A few problems in latest changes, and some benchmark findings...

patriknw commented:
Closed due to inactivity and more difficulties than anticipated.
I have prepared the config structure for adding lz4 later in https://github.com/akka/akka/pull/27889
